Add masking strategies to message transforms

supreethmanyam opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [x] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
https://github.com/pytorch/torchtune/issues/2261

#### Changelog
What are the changes made in this PR?
* Added `mask_messages` utility in `torchtune/data/_messages.py` to mask each message given a list of messages based on user provided masking_strategy
* Fixes in-built message transforms to always mask system message
* Added tests for calling message transforms with masking strategies

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

supreethmanyam commented:
@RdoubleA Added `mask_messages` utility in `torchtune/data/_messages.py` along with relevant unit tests.
I will update the documentation once the changes look good.

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -176,12 +185,24 @@ class InputOutputToMessages(Transform):
     def __init__(
         self,
-        train_on_input: bool = False,
+        masking_strategy: Optional[str] = "train_on_all",
```
if the default value for `train_on_input` was False, this would correspond to `train_on_assistant`. Let's keep that same default behavior.

supreethmanyam replied:
Updated default strategy

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -305,11 +332,23 @@ class ChosenRejectedToMessages(Transform):
     def __init__(
         self,
-        train_on_input: bool = False,
+        masking_strategy: Optional[str] = "train_on_all",
```
same comment here

supreethmanyam replied:
Updated default strategy

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -406,20 +445,35 @@ class ShareGPTToMessages(Transform):
     def __init__(
         self,
-        train_on_input: bool = False,
+        masking_strategy: Optional[str] = "train_on_all",
```
same comment here

supreethmanyam replied:
Updated default strategy

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -623,55 +691,12 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
         return {"messages": updated_messages}
 
 
-def validate_messages(
```
not sure if this is still used but let's keep this here for now

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -682,17 +707,38 @@ class AlpacaToMessages(Transform):
     def __init__(
-        self, train_on_input: bool = True, column_map: Optional[Dict[str, str]] = None
+        self,
+        masking_strategy: Optional[str] = "train_on_all",
```
contrary to the other comment, this default value should remain `train_on_all` since `train_on_input` was defaulted to True. so let's keep this

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -726,14 +772,94 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        ValueError: If the masking strategy is not one of the supported strategies:
+                    `train_on_all`, `train_on_assistant`, `train_on_last`.
+    """
+    if masking_strategy not in ["train_on_all", "train_on_assistant", "train_on_last"]:
```
One suggestion that I didn't mention in the issue is that we could make `masking_strategy` an Enum so that it can be type checked more easily. users can still specify as a string in the config and in the message transforms above, but here in `mask_messages` you could convert to the Enum, and the ValueError will automatically be captured.

supreethmanyam replied:
Converted masking_strategy to Enum and raises ValueErroe automatically if doesn't match with mapping

RdoubleA reviewed (commented):
This looks great, really like the tests 👌 left a couple of comments

supreethmanyam commented:
@RdoubleA 

In addition to converting masking_strategy to Enum and fixing default strategy in built-in transforms, I made following small changes:
1. I modified structure of `column_map` usage in `InputOutputToMessages` and `AlpacaToMessages` to make it consistent across transforms
2. I realized there were no unit tests for `AlpacaToMessages`, I added a few tests.

ebsmothers commented on torchtune/data/_messages.py:
```diff
@@ -177,31 +193,42 @@ class InputOutputToMessages(Transform):
     def __init__(
         self,
-        train_on_input: bool = False,
+        masking_strategy: Optional[str] = "train_on_assistant",
```
Why do we switch the order of these arguments? If someone is passing `InputOutputToMessages(False)` this will break them, right? (Since we unfortunately don't enforce keyword-only args on these APIs). Personally I would leave train_on_input in the first position and put masking_strategy last

RdoubleA replied:
agreed, let's keep the positions of the arguments the same until we deprecate

ebsmothers commented on tests/torchtune/data/test_messages.py:
```diff
@@ -267,6 +296,55 @@ def test_call_train_on_input(self, sample):
+    @pytest.mark.parametrize(
+        "masking_strategy, expected_masks",
+        [
+            ("train_on_all", [[False, False], [False, False]]),
+            ("train_on_assistant", [[True, False], [True, False]]),
+            ("train_on_last", [[True, False], [True, False]]),
```
Ideally we should explicitly test a case where the results of `train_on_assistant` and `train_on_last` differ

supreethmanyam replied:
I have added test cases using multi turn samples where results `train_on_assistant` and `train_on_last` differ where applicable (ChosenRejectedToMessages, ShareGPTToMessages, OpenAIToMessages)

ebsmothers commented on tests/torchtune/data/test_messages.py:
```diff
@@ -479,6 +609,200 @@ def test_call_image_messages(self, mock_load_image):
         mock_load_image.assert_called_once_with("https://example.com")
 
 
+class TestAlpacaToMessages:
```
What's the difference between this and the existing `TestAlpacaToMessages` [here](https://github.com/pytorch/torchtune/blob/67a8706abd993d4b03c70506075a2a9804919148/tests/torchtune/datasets/test_alpaca_dataset.py#L93)? They look fairly similar

supreethmanyam replied:
I did not know we had `TestAlpacaToMessages` in test_alpaca_dataset.py and added one in test_messages.py. 

Both are similar and the one in test_messages.py runs more tests. Please let me know if I can remove TestAlpacaToMessages from [here](https://github.com/pytorch/torchtune/blob/67a8706abd993d4b03c70506075a2a9804919148/tests/torchtune/datasets/test_alpaca_dataset.py#L93)

ebsmothers commented on torchtune/data/_messages.py:
```diff
@@ -709,32 +782,102 @@ def __init__(
         return {"messages": messages}
+
+
+def validate_messages(
```
Similar q here: don't we have an existing `validate_messages` function?

RdoubleA replied:
it was just moved down here I believe

supreethmanyam replied:
I moved `validate_messages` down to keep all the message transforms at one place.

ebsmothers reviewed (commented):
Sorry @supreethmanyam, I think this PR fell through the cracks a bit. I have a couple comments but after that I think it looks good. cc @RdoubleA to confirm

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -683,18 +714,60 @@ class AlpacaToMessages(Transform):
     ):
-        self.train_on_input = train_on_input
-        self.column_map = column_map
+        if train_on_input is not None:
```
I thought we had a deprecated param utility, but it looks like it never got reviewed: https://github.com/pytorch/torchtune/pull/2321. should make a note there to also replace this code if this PR merges first @ebsmothers

supreethmanyam commented:
@RdoubleA @ebsmothers 
I have made all the changes and updated the documentation. Requesting another review.

Current references to `train_on_input` in the docs comes from dataset APIs, hence I haven't made any changes yet. Please let me know if I should create another issue for dataset API changes. I will be happy to make changes to dataset APIs to reflect using `masking_strategy` instead of `train_on_input` and update docs accordingly.

ebsmothers commented:
Thanks @supreethmanyam for your patience on this one. Since we have not been great about timely reviews here, I took the liberty of pushing some changes directly to the PR (hope that's alright). Mainly I changed the deprecation logic slightly: when `train_on_input` and `masking_strategy` are both passed, we should still use `train_on_input` but warn the user (otherwise it's possible for silent behavior changes without the user actually migrating). While doing this, I noticed that our SlimOrca test case seems incorrect -- it appears as though we were never masking the system message (see [here](https://github.com/pytorch/torchtune/blob/9b05ac46ded4424bcecf3984ace42704af6a7422/torchtune/data/_messages.py#L490-L492)). cc @RdoubleA to sanity check me here. After this (and green CI) I think this should be good to land. Thanks again for your patience!

RdoubleA commented:
That's a tough one... it does look like system is being masked, but if train_on_input is true it will not be masked, same as the user message. So yeah, big oof

ebsmothers reviewed (approved):
Really appreciate you adding this @supreethmanyam!
